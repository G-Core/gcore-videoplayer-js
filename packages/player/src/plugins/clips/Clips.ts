import { Container, Events, UICorePlugin, $, template } from '@clappr/core'
import { trace } from '@gcorevideo/utils'
import assert from 'assert'

import { TimeProgress, TimeValue } from '../../playback.types.js'
import type { ZeptoResult } from '../../types.js'
import '../../../assets/clips/clips.scss'
import { ClipDesc } from './types.js'
import { buildSvg, parseClips } from './utils.js'
import clipsHTML from '../../../assets/clips/clips.ejs'

const T = 'plugins.clips'

/**
 * Configuration options for the {@link Clips} plugin.
 * @beta
 */
export interface ClipsPluginSettings {
  /**
   * The compiled text of the clips description, one clip per line in format  :
   * `HH:MM:SS text` or `MM:SS text` or `SS text`
   */
  text: string
}

const VERSION = '2.22.16'
const CLAPPR_VERSION = '0.11.4'

/**
 * `PLUGIN` that allows marking up the timeline of the video
 * @beta
 * @remarks
 * The plugin decorates the seekbar with notches to indicate the clips of the video and displays current clip text in the left panel
 *
 * Depends on:
 *
 * - {@link MediaControl}
 *
 * Configuration options - {@link ClipsPluginSettings}
 */
export class Clips extends UICorePlugin {
  private barStyle: HTMLStyleElement | null = null

  private clips: ClipDesc[] = []

  private oldContainer: Container | undefined

  private svgMask: ZeptoResult | null = null

  private static readonly template = template(clipsHTML)

  /**
   * @internal
   */
  get name() {
    return 'clips'
  }

  /**
   * @internal
   */
  override get attributes() {
    return {
      class: 'media-control-clips',
    }
  }

  get version() {
    return VERSION
  }

  get supportedVersion() {
    return { min: CLAPPR_VERSION }
  }

  /**
   * @internal
   */
  override bindEvents() {
    this.listenToOnce(this.core, Events.CORE_READY, this.onCoreReady)
    this.listenTo(this.core, Events.CORE_RESIZE, this.playerResize)
    this.listenTo(
      this.core,
      Events.CORE_ACTIVE_CONTAINER_CHANGED,
      this.onContainerChanged,
    )
  }

  override render() {
    trace(`${T} render`)
    if (!this.options.clips) {
      return this
    }
    this.$el.html(Clips.template())
    this.$el.hide()
    return this
  }

  override destroy() {
    if (this.barStyle) {
      this.barStyle.remove()
      this.barStyle = null
    }
    return super.destroy()
  }

  override disable() {
    if (this.barStyle) {
      this.barStyle.remove()
      this.barStyle = null
    }
    return super.disable()
  }

  override enable() {
    this.render()
    return super.enable()
  }

  /**
   * Get the text of the clip at the given time
   * @param time - The time to get the text for
   * @returns The text of the clip at the given time
   */
  getText(time: TimeValue): string | undefined {
    return this.clips.find((clip) => clip.start <= time && clip.end >= time)?.text
  }

  private onCoreReady() {
    trace(`${T} onCoreReady`)
    const mediaControl = this.core.getPlugin('media_control')
    assert(mediaControl, 'media_control plugin is required')

    this.parseClips(this.options.clips.text)
    this.listenTo(mediaControl, Events.MEDIACONTROL_RENDERED, this.onMcRender)
  }

  private onMcRender() {
    trace(`${T} onMcRender`)
    const mediaControl = this.core.getPlugin('media_control')
    mediaControl.mount('clips', this.$el)
  }

  private onContainerChanged() {
    trace(`${T} onContainerChanged`)
    // TODO figure out the conditions of changing the container (without destroying the previous one)
    if (this.oldContainer) {
      this.stopListening(
        this.oldContainer,
        Events.CONTAINER_TIMEUPDATE,
        this.onTimeUpdate,
      )
    }
    this.oldContainer = this.core.activeContainer
    if (this.svgMask) {
      this.svgMask.remove()
      this.svgMask = null
    }
    this.listenTo(
      this.core.activeContainer,
      Events.CONTAINER_TIMEUPDATE,
      this.onTimeUpdate,
    )
  }

  private playerResize() {
    const duration = this.core.activeContainer.getDuration()
    if (duration) {
      this.makeSvg(duration)
    }
  }

  private onTimeUpdate(event: TimeProgress) {
    if (!this.svgMask) {
      this.makeSvg(event.total)
    }
    for (const value of this.clips) {
      if (
        (event.current >= value.start && !value.end) ||
        event.current < value.end
      ) {
        this.setClipText(value.text)
        break
      }
    }
  }

  private parseClips(text: string) {
    this.clips = parseClips(text)
  }

  private makeSvg(duration: number) {
    const svg = buildSvg(
      this.clips,
      duration,
      this.core.activeContainer.$el.width(),
    )
    this.setSVGMask(svg)
  }

  private setSVGMask(svg: string) {
    if (this.svgMask) {
      this.svgMask.remove()
    }

    this.svgMask = $(svg)
    this.$el.append(this.svgMask)
    if (!this.barStyle) {
      this.barStyle = document.createElement('style')
      this.barStyle.textContent = `
.bar-container[data-seekbar] {
  clip-path: url("#myClip");
}`
      this.$el.append(this.barStyle)
    }
  }

  private setClipText(text: string) {
    if (text) {
      this.$el.show().find('#clips-text').text(text)
    } else {
      this.$el.hide()
    }
  }
}
