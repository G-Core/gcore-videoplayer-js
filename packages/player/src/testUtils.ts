import { $, UICorePlugin } from '@clappr/core'
import Events from 'eventemitter3'
import { vi } from 'vitest'

export function createMockCore(
  options: Record<string, unknown> = {},
  container: any = createMockContainer(options),
) {
  const el = document.createElement('div')
  const emitter = new Events()
  return Object.assign(emitter, {
    el,
    $el: $(el),
    activePlayback: container.playback,
    activeContainer: container,
    i18n: {
      t: vi.fn().mockImplementation((key: string) => key),
    },
    options: {
      ...options,
    },
    configure: vi.fn(),
    getPlaybackType: vi.fn(),
    getPlugin: vi.fn(),
    load: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockPlugin() {
  return Object.assign(new Events(), {
    enable: vi.fn(),
    disable: vi.fn(),
  })
}

export function createSpinnerPlugin() {
  return Object.assign(createMockPlugin(), {
    show: vi.fn(),
    hide: vi.fn(),
  })
}

export function createMockPlayback(name = 'mock') {
  const emitter = new Events()
  return Object.assign(emitter, {
    name,
    currentLevel: -1,
    el: document.createElement('video'),
    dvrEnabled: false,
    dvrInUse: false,
    isAudioOnly: false,
    levels: [],
    consent: vi.fn(),
    play: vi.fn(),
    pause: vi.fn(),
    stop: vi.fn(),
    destroy: vi.fn(),
    seek: vi.fn(),
    seekPercentage: vi.fn(),
    getDuration: vi.fn().mockImplementation(() => 100),
    enterPiP: vi.fn(),
    exitPiP: vi.fn(),
    getPlaybackType: vi.fn(),
    getStartTimeOffset: vi.fn().mockImplementation(() => 0),
    getCurrentTime: vi.fn().mockImplementation(() => 0),
    isHighDefinitionInUse: vi.fn().mockImplementation(() => false),
    mute: vi.fn(),
    unmute: vi.fn(),
    volume: vi.fn(),
    configure: vi.fn(),
    attemptAutoPlay: vi.fn().mockImplementation(() => true),
    canAutoPlay: vi.fn().mockImplementation(() => true),
    onResize: vi.fn().mockImplementation(() => true),
    setPlaybackRate: vi.fn(),
    switchAudioTrack: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockContainer(
  options: Record<string, unknown> = {},
  playback: any = createMockPlayback(),
) {
  const el = playback.el
  const emitter = new Events()
  return Object.assign(emitter, {
    el,
    playback,
    options: {
      ...options,
    },
    $el: $(el),
    disableMediaControl: vi.fn(),
    enableMediaControl: vi.fn(),
    enterPiP: vi.fn(),
    exitPiP: vi.fn(),
    getDuration: vi.fn().mockReturnValue(0),
    getPlugin: vi.fn(),
    getPlaybackType: vi.fn(),
    getStartTimeOffset: vi.fn().mockReturnValue(0),
    isDvrInUse: vi.fn().mockReturnValue(false),
    isDvrEnabled: vi.fn().mockReturnValue(false),
    isHighDefinitionInUse: vi.fn().mockReturnValue(false),
    isPlaying: vi.fn().mockReturnValue(false),
    play: vi.fn(),
    seek: vi.fn(),
    seekPercentage: vi.fn(),
    switchAudioTrack: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockMediaControl(core: any) {
  const mediaControl = new UICorePlugin(core)
  // TODO <div class="media-control-layer">
  mediaControl.$el.html(
    `<div class="media-control-left-panel" data-media-control></div>
    <div class="media-control-right-panel" data-media-control></div>
    <div class="media-control-center-panel" data-media-control></div>`,
  )
  // @ts-ignore
  mediaControl.mount = vi.fn()
  // @ts-ignore
  mediaControl.container = core.activeContainer
  // @ts-ignore
  mediaControl.toggleElement = vi.fn()
  vi.spyOn(mediaControl, 'trigger')
  core.$el.append(mediaControl.$el)
  return mediaControl
}

export function createMockBottomGear(core: any) {
  const plugin: any = new UICorePlugin(core)
  plugin.getItem = vi.fn()
  plugin.addItem = vi.fn().mockImplementation((name: string, $el: any) => {
    const existing = plugin.$el.find(`[data-${name}]`)
    if (existing.length) {
      return existing
    }
    return $('<li></li>')
      .attr(`data-${name}`, '')
      .append($el)
      .appendTo(plugin.$el)
  })
  plugin.refresh = vi.fn()
  return plugin
}
