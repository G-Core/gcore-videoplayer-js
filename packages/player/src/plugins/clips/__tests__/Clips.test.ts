import { beforeEach, describe, expect, it, vi } from 'vitest'
import { Clips } from '../Clips'
import { createMockCore, createMockMediaControl } from '../../../testUtils'
import { Events } from '@clappr/core'

// import { LogTracer, Logger, setTracer } from '@gcorevideo/utils'

// Logger.enable('*')
// setTracer(new LogTracer('Clips.text'))

describe('Clips', () => {
  let core: any
  let mediaControl: any
  let clips: Clips
  beforeEach(() => {
    core = createMockCore({
      clips: {
        text: `
      00:00:00 Introduction
      00:05:00 Main part
      00:15:00 Conclusion
      `,
      },
    })
    mediaControl = createMockMediaControl(core)
    core.getPlugin.mockImplementation((name: string) => {
      if (name === 'media_control') return mediaControl
      return null
    })
    clips = new Clips(core)
    core.emit(Events.CORE_READY)
    core.emit(Events.CORE_ACTIVE_CONTAINER_CHANGED, core.activeContainer)
    vi.spyOn(core.activeContainer.$el, 'width').mockReturnValue(600)
    core.activeContainer.emit(Events.CONTAINER_TIMEUPDATE, {
      current: 0,
      total: 1200,
    })
  })
  it('should render indicator', () => {
    expect(clips.el.innerHTML).toMatchSnapshot()
  })
  it('should render notches on the seek bar', () => {
    const svg = clips.$el.find('svg')
    expect(svg).toBeDefined()
    expect(svg?.find('rect').length).toBe(3)
  })
  describe('as time progresses', () => {
    describe.each([
      [60, 'Introduction'],
      [310, 'Main part'],
      [1001, 'Conclusion'],
    ])('@%s', (time, expected) => {
      beforeEach(() => {
        core.activeContainer.emit(Events.CONTAINER_TIMEUPDATE, {
          current: time,
          total: 1200,
        })
      })
      it(`text should be "${expected}"`, () => {
        expect(clips.$el.find('#clips-text').text()).toBe(expected)
      })
    })
  })
  describe('when media control is rendered', () => {
    beforeEach(() => {
      mediaControl.trigger(Events.MEDIACONTROL_RENDERED)
    })
    it('should mount the indicator', () => {
      expect(mediaControl.slot).toHaveBeenCalledWith('clips', clips.$el)
    })
  })
})
