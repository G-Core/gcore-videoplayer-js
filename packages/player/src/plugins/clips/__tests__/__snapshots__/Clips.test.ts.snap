// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ClipsPlugin > should render indicator 1`] = `
"<div class="media-clip-text" id="clips-text">Introduction</div><svg width="0" height="0">
<defs>
<clipPath id="myClip">
<rect x="0" y="0" width="148" height="30"></rect>
<rect x="150" y="0" width="298" height="30"></rect>
<rect x="450" y="0" width="148" height="30"></rect>
</clipPath></defs></svg><style>
.bar-container[data-seekbar] {
  clip-path: url("#myClip");
}</style>"
`;
